### PR TITLE
Support for Full Elasticsearch DSL

### PR DESCRIPTION
**Pull request by gingerwizard**
This pull request adds full support for the Elasticsearch DSL.  This allows us to query elasticsearch with more than just the query_string, utilising the full syntax.

The query is required to specify the DSL as a template, which gets populated from the event fields e.g.

```
elasticsearch {
          hosts => ["localhost:9200"]
          query_template => "/opt/logstash/logstash-5.0.0-alpha4/config/ip_address_query.json"
          index => "threats-2016.07.26"
          "fields" => [ ["source", "threat_source"],["type", "threat_type"],["direction", "threat_direction"] ]
   }
```

Example template:

```
{
  "query": {
    "percolate": {
      "field" : "query",
      "document_type" : "web_access",
      "document": {
        "ip_address": "%{message}"
      }
    }
  },
  "size": 1
}
```

As shown above this means we can now enrich documents at index time using percolator, as the new 5.0 syntax is just a [query](https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-percolate-query.html)

### CODE REVIEW

**Comment by gingerwizard** [comment]
@ph @jordansissel 
Thoughts appreciated.

**Comment by ph** [comment]
I've been discussing with @gingerwizard and I really like the idea to allow people to use directly the query DSL, I think it opens the plugin to a new broad of usages.

From what I see in the plugin we have changed the option `query` to `query_string`, this change make this plugin backward incompatible with any previous releases.

Personally, I believe that `query_string` is a better name for it, but we should the change more gracefully.  We could mark the `query` as deprecated, add the new `query_string` but allow people to use one or the other before we make the final removal. WDYT?

File: logstash-filter-elasticsearch.gemspec [hunk 1]
```diff
@@ -1,7 +1,7 @@
 Gem::Specification.new do |s|
 
   s.name            = 'logstash-filter-elasticsearch'
-  s.version         = '3.0.2'
+  s.version         = '3.1.0'
```
Comment (ph) [inline logstash-filter-elasticsearch.gemspec]: even if we keep the old option, this is still worth a minor bump! good choice.

**Comment by ph** [comment]
Two things are also missing:
- update the changelog.md
- Maybe add a complete usage in the head of the plugin so it will show up in the documentation.

**Comment by gingerwizard** [comment]
Agreed @ph re the naming of query to query_string. I'll make all the requested changes + an integration test.

**Comment by samwoods1** [comment]
I'd love to see this PR moved forward.  I'd be happy to take on making the suggested changes and creating an integration test if this is stalled? Thanks!  I'm also already thinking of adding support for aggregate values as well, but that would require a few more updates and can be a separate PR.

**Comment by gingerwizard** [comment]
@ph completely forgot about this! soo sorry! Did the changes and made another commit. Please review.

File: lib/logstash/filters/elasticsearch.rb [hunk 2]
```diff
@@ -38,9 +73,13 @@ class LogStash::Filters::Elasticsearch < LogStash::Filters::Base
   # Comma-delimited list of index names to search; use `_all` or empty string to perform the operation on all indices
   config :index, :validate => :string, :default => ""
 
-  # Elasticsearch query string. Read the Elasticsearch query string documentation
+  # Elasticsearch query string. Read the Elasticsearch query string documentation. Deprecated.
   # for more info at: https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-query-string-query.html#query-string-syntax
-  config :query, :validate => :string
+  config :query, :validate => :string, :deprecated => true
```
Comment (ph) [inline lib/logstash/filters/elasticsearch.rb]: I've had some though about our discussion, I think we don't really need to mark the query as deprecated yet, I think it's handy for easy query. If you need something more complex use the `query_template`

File: lib/logstash/filters/elasticsearch.rb [hunk 3]
```diff
@@ -38,9 +73,13 @@ class LogStash::Filters::Elasticsearch < LogStash::Filters::Base
...
+
+  # File path to elasticsearch query in DSL format. Read the Elasticsearch query documentation
+  # for more info at: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html
+  config :query_template, :validate => :string
```
Comment (ph) [inline lib/logstash/filters/elasticsearch.rb]: Add a note in the documentation to tell people that it support the fieldref syntax.

**Comment by ph** [comment]
Simple changes, lets make sure we rebase this @gingerwizard than I will merge it. good work!

**Comment by gingerwizard** [comment]
@ph changes applied. Can we merge?

**Comment by andreyrusanov** [comment]
@ph @gingerwizard 
sorry for bugging, but any progress on this PR? This functionality is something I'm really looking forward!(unfortunately I don't know Ruby so can't participate)

**Comment by gingerwizard** [comment]
@andreyrusanov i think this is just waiting to be merged. @ph do we need anything else?

**Comment by andreyrusanov** [comment]
@gingerwizard at least there is merge conflicts, so maybe you need to pull, resolve them and push back?

**Comment by andreyrusanov** [comment]
@gingerwizard @ph 
sorry for bothering, guys, could you put some efforts on the PR?

**Comment by ph** [comment]
LGTM

**Comment by ph** [comment]
merged in #53

**Comment by ph** [comment]
v3.1.0 is now on rubygems!

**Pull request closed without merging**